File: IrrgartenP2/IrrgartenJava/src/irrgartenjava/Game.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgartenjava;

/**
 *
 * @author moham
 */// Game.java
import java.util.ArrayList;

public class Game {
    private Labyrinth labyrinth;
    private ArrayList<Player> players;
    private ArrayList<Monster> monsters;
    private Player currentPlayer;
    private int currentPlayerIdx;
    private String log;

    public Game(int nPlayers, int nRows, int nCols) {
        this.labyrinth = new Labyrinth(nRows, nCols);
        this.players = new ArrayList<>();
        this.monsters = new ArrayList<>();
        this.currentPlayerIdx = Dice.whoStarts(nPlayers);
        this.log = "";

        // Create players
        for (int i = 0; i < nPlayers; i++) {
            players.add(new Player(i, Dice.randomIntelligence(), Dice.randomStrength()));
        }
        this.currentPlayer = players.get(currentPlayerIdx);

        configureLabyrinth();
    }

    public boolean finished() {
        return labyrinth.haveAWinner();
    }

    public GameState getGameState() {
        return new GameState(labyrinth.toString(), players.toString(), monsters.toString(), currentPlayerIdx, finished(), log);
    }

    public void configureLabyrinth() {
        // Add exit to a random empty position
        int[] exitPos = labyrinth.randomEmptyPos();
        labyrinth.setExit(exitPos[0], exitPos[1]);

        // Add monsters to random positions
        for (int i = 0; i < players.size(); i++) {
            Monster monster = new Monster(Dice.randomStrength(), Dice.randomIntelligence());
            int[] pos = labyrinth.randomEmptyPos();
            labyrinth.addMonster(pos[0], pos[1], monster);
            monsters.add(monster);
        }

        // Spread players in random positions
        labyrinth.spreadPlayers(players);
    }

    public void nextPlayer() {
        currentPlayerIdx = (currentPlayerIdx + 1) % players.size();
        currentPlayer = players.get(currentPlayerIdx);
    }

    public void logPlayerWon() {
        log += "Player " + currentPlayer + " has reached the exit and won the game!\n";
    }

    public void logMonsterWon() {
        log += "A monster has defeated player " + currentPlayer + ".\n";
    }

    public void logResurrected() {
        log += "Player " + currentPlayer + " has been resurrected.\n";
    }

    public void logPlayerSkipTurn() {
        log += "Player " + currentPlayer + " skips their turn.\n";
    }

    public void logPlayerNoOrders() {
        log += "Player " + currentPlayer + " received no valid move orders.\n";
    }

    public void logNoMonster() {
        log += "No monster present for combat.\n";
    }

    public void logRounds(int rounds, int max) {
        log += "Combat ended after " + rounds + " of " + max + " rounds.\n";
    }

    public boolean nextStep(Directions preferredDirection) {
        throw new UnsupportedOperationException("Method not implemented yet.");
    }

    public Directions actualDirection(Directions preferredDirection) {
        throw new UnsupportedOperationException("Method not implemented yet.");
    }

    public GameCharacter combat(Monster monster) {
        throw new UnsupportedOperationException("Method not implemented yet.");
    }

    public void manageReward(GameCharacter winner) {
        throw new UnsupportedOperationException("Method not implemented yet.");
    }

    public void manageResurrection() {
        throw new UnsupportedOperationException("Method not implemented yet.");
    }
}

File: IrrgartenP2/irrgartenRuby/game.rb
# game.rb
class Game
  attr_reader :labyrinth, :players, :monsters, :current_player, :log

  ROW = 0
  COL = 1 #Quitar despues de configurar el configurarLabyrinth


  def initialize(n_players, n_rows, n_cols)
    @labyrinth = Labyrinth.new(n_rows, n_cols)
    @players = Array.new(n_players) { |i| Player.new(i, Dice.random_intelligence, Dice.random_strength) }
    @monsters = []
    @current_player_idx = Dice.who_starts(n_players)
    @current_player = @players[@current_player_idx]
    @log = ""

    #configure_labyrinth
  end

  def finished?
    @labyrinth.have_a_winner?
  end

  def get_game_state
    {
      labyrinth: @labyrinth.to_s,
      players: @players.map(&:to_s),
      monsters: @monsters.map(&:to_s),
      current_player_idx: @current_player_idx,
      finished: finished?,
      log: @log
    }
  end

  def configure_labyrinth
    exit_pos = @labyrinth.random_empty_pos
    @labyrinth.set_exit(exit_pos[ROW], exit_pos[COL])

    @players.size.times do
      monster = Monster.new(Dice.random_strength, Dice.random_intelligence)
      monster_pos = @labyrinth.random_empty_pos
      @labyrinth.add_monster(monster_pos[ROW], monster_pos[COL], monster)
      @monsters << monster
    end
  end

  def next_player
    @current_player_idx = (@current_player_idx + 1) % @players.size
    @current_player = @players[@current_player_idx]
  end

  def log_player_won
    @log += "Player #{@current_player} has reached the exit and won the game!\n"
  end

  def log_monster_won
    @log += "A monster has defeated player #{@current_player}.\n"
  end

  def log_resurrected
    @log += "Player #{@current_player} has been resurrected.\n"
  end

  def log_player_skip_turn
    @log += "Player #{@current_player} skips their turn.\n"
  end

  def log_player_no_orders
    @log += "Player #{@current_player} received no valid move orders.\n"
  end

  def log_no_monster
    @log += "No monster present for combat.\n"
  end

  def log_rounds(rounds, max)
    @log += "Combat ended after #{rounds} of #{max} rounds.\n"
  end

  def next_step(preferred_direction)
    raise NotImplementedError, "Method not implemented yet."
  end

  def actual_direction(preferred_direction)
    raise NotImplementedError, "Method not implemented yet."
  end

  def combat(monster)
    raise NotImplementedError, "Method not implemented yet."
  end

  def manage_reward(winner)
    raise NotImplementedError, "Method not implemented yet."
  end

  def manage_resurrection
    raise NotImplementedError, "Method not implemented yet."
  end
end
